File: js/__tests__/hash_table.test.js
// 散列

const HashTable = require('../data_structures/hash_table')

describe('HashTable', () => {

    it('put', () => {
        const hashTable = new HashTable()
        hashTable.put('hello')
        // console.log(hashTable.table)
        // expect(hashTable.table).toContain('hello')
        expect(hashTable.table[121]).toContain('hello')
    })

    it('get', () => {
        const hashTable = new HashTable();
        const worldKey = hashTable.put('world');
        expect(hashTable.get(worldKey)).toContain('world')
    })

    it('conflict', () => {
        // 散列需要处理碰撞问题
        const hashTable = new HashTable();
        const helloKey = hashTable.put('hello')
        const noHaKey = hashTable.put('lloeh')
        const otherKey = hashTable.put('bear')
        expect(helloKey).toEqual(noHaKey)
        expect(hashTable.table[helloKey]).toContain('hello')
        expect(hashTable.table[helloKey]).toContain('lloeh')
        expect(hashTable.table[otherKey]).toContain('bear')
        // console.log(hashTable.table[helloKey])
        // console.log(hashTable.table[otherKey])
    })

})

File: py/__tests__/hash_table_test.py
import unittest
from data_structures.hash_table import HashTable


class TestHashTable(unittest.TestCase):
    def test_put(self):
        hash_table = HashTable()
        hash_key = hash_table.put('hello')
        self.assertEqual(hash_table.table[hash_key], 'hello')

    def test_conflict(self):
        hash_table = HashTable()
        hello_key = hash_table.put('helloa')
        llohe_key = hash_table.put('llohae')
        self.assertEqual(hello_key + 1, llohe_key)
        # hellob 的hash_key 为 helloa + 1, 会与已经产生过碰撞的llohae再次冲突
        db_conf_key = hash_table.put('hellob')
        self.assertEqual(db_conf_key, hello_key + 2)
        self.assertEqual(db_conf_key, llohe_key + 1)
        self.assertEqual(hash_table.table[db_conf_key], 'hellob')
